### DISCUSSION TITLE
draw_rect scaled to very small size

### DISCUSSION THREAD

**Discussion by marvinamhere** [pymupdf/PyMuPDF#3491] [Q&A]
### Description of the bug

draw_rect behave different on existing page and new page

### How to reproduce the bug

i have an existing pdf file and when i use draw_rect to draw on one of its page , eg , (100,100,600,600) , but it scaled to (20,20,200,200) , the rectangle is very small . even i set the width to more like 3000, it only can occupy 1/4 area of the whole page on  left-top area .  
but when create a new page , and draw on it , it works fine . 

my question is , is there some possible reason that make draw_rect such behavior . actually not only draw_rect ,but also  insert_text also this behavior .  and i compared the new page and existing page , the transformation and other matrix are all same . i think the existing pdf page must have difference with new page , but i can't find it ...

### PyMuPDF version

1.24.2

### Operating system

MacOS

### Python version

3.12

**Comment by JorjMcKie** [comment]
We can only accept a bug report complete with an example file. I also noticed that you do not use the current version.

**Comment by marvinamhere** [comment]
hi ,how to remove all CTM from the pdf page ? i saw some "cm" instruction left in existing page  : (0.24,0,0,-0.24,0,840) and when i darw_rect it will mulitple that ctm  with input rect.  
or how to get the current CTM value from current page ? after get it , how to get a reverse one as the Morph ?

**Comment by JorjMcKie** [comment]
> hi ,how to remove all CTM from the pdf page ? i saw some "cm" instruction left in existing page : (0.24,0,0,-0.24,0,840) and when i darw_rect it will mulitple that ctm with input rect. or how to get the current CTM value from current page ? after get it , how to get a reverse one as the Morph ?

This consideration will not resolve your problem. Please repeat using the current PyMuPDF version 1.24.3.

**Comment by marvinamhere** [comment]
i am using 1.24.4 to reproduce this . 
i know the previous cm is (0.24,0,0,-0.24,0,0 )
how to pass the morph param to generate 
"4.16,0,0,4.16,0,0 cm" (the inverse of above cm ) instruction ? (thus before draw , the two matrix multipy equal to 1)

from below code seems it depends the start point and a matrix , how to decide them ?


in shape.finish there is : 

> if pymupdf.CheckMorph(morph):
            m1 = pymupdf.Matrix(
                1, 0, 0, 1, morph[0].x + self.x, self.height - morph[0].y - self.y
            )
            mat = ~m1 * morph[1] * m1
            self.draw_cont = _format_g(pymupdf.JM_TUPLE(mat)) + " cm\n" + self.draw_cont

**Comment by JorjMcKie** [comment]
First of all, this seems to be no bug report any longer.
I am therefore going to move this thread to the Discussions tab. We can pick up questions there.

**Comment by JorjMcKie** [comment]
The morph parameter executes a matrix ... **_any_** matrix with `matrix.e == matrix.f == 0`. So you can scale, rotate, tilt, flip, whatever you like.
You also must provide a point that should remain fixed under the matrix execution. E.g. for a rotation this obviously the point around which to rotate. If the matrix is a flipping, then a full line containing that point will also be retained.

So you must first answer the question what the fixpoint should be. Then it looks like you want to scale, so use `matrix=pymupdf.Matrix(scale, scale)` ... with the desired `scale` value.
Then specify `morph=(fixpoint, matrix)`.

**Comment by marvinamhere** [comment]
thanks , sorry , the issue suddenly disappear...my bad ..even i dont know where i bad it ..i tried two days ..thanks
